mlir/lib/Conversion/ArithToAPFloat/ArithToAPFloat.cpp

```diff
@@ -0,0 +1,120 @@
+//===- ArithToAPFloat.cpp - Arithmetic to APFloat Conversion --------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Conversion/ArithToAPFloat/ArithToAPFloat.h"
+
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Arith/Transforms/Passes.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Func/Utils/Utils.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/IR/Verifier.h"
+#include "mlir/Transforms/WalkPatternRewriteDriver.h"
+
+namespace mlir {
+#define GEN_PASS_DEF_ARITHTOAPFLOATCONVERSIONPASS
+#include "mlir/Conversion/Passes.h.inc"
+} // namespace mlir
+
+using namespace mlir;
+using namespace mlir::func;
+
+/// Helper function to look up or create the symbol for a runtime library
+/// function for a binary arithmetic operation.
+///
+/// Parameter 1: APFloat semantics
+/// Parameter 2: Left-hand side operand
+/// Parameter 3: Right-hand side operand
+///
+/// This function will return a failure if the function is found but has an
+/// unexpected signature.
+///
+static FailureOr<Operation *>
+lookupOrCreateBinaryFn(OpBuilder &b, Operation *moduleOp, StringRef name,
+                       SymbolTableCollection *symbolTables = nullptr) {
+  auto i32Type = IntegerType::get(moduleOp->getContext(), 32);
+  auto i64Type = IntegerType::get(moduleOp->getContext(), 64);
+  return lookupOrCreateFnDecl(b, moduleOp,
+                              (llvm::Twine("_mlir_apfloat_") + name).str(),
+                              {i32Type, i64Type, i64Type}, {i64Type},
+                              /*setPrivate=*/true, symbolTables);
+}
+
+/// Rewrite a binary arithmetic operation to an APFloat function call.
+template <typename OpTy, const char *APFloatName>
+struct ArithOpToAPFloatConversion final : OpRewritePattern<OpTy> {
+  using OpRewritePattern<OpTy>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(OpTy op,
+                                PatternRewriter &rewriter) const override {
+    auto moduleOp = op->template getParentOfType<ModuleOp>();
+    if (!moduleOp) {
+      return rewriter.notifyMatchFailure(
+          op, "arith op must be contained within a builtin.module");
+    }
+    // Get APFloat function from runtime library.
+    FailureOr<Operation *> fn =
+        lookupOrCreateBinaryFn(rewriter, moduleOp, APFloatName);
+    if (failed(fn))
+      return op->emitError("failed to lookup or create APFloat function");
```

Comment:
Could we do something better here? E.g., its not an error of this op if we got here I believe, it means we had mismatching signatures/overlapping names, and could be useful to report the mismatched signature.